spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,296 @@
+package org.apache.spark.sql.cassandra
+
+import org.apache.spark.Logging
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.cassandra.CassandraSQLRow.CassandraSQLRowReader
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+
+import com.datastax.spark.connector._
+import com.datastax.spark.connector.cql.{CassandraConnector, ColumnDef}
+import com.datastax.spark.connector.rdd.ReadConf
+import com.datastax.spark.connector.writer.{WriteConf, SqlRowWriter}
+
+import DataTypeConverter._
+
+/**
+ *  Implements [[BaseRelation]] and [[InsertableRelation]].
+ *  It inserts data into Cassandra table. It also provides some
+ *  helper methods and variables for sub class to use.
+ */
+private[cassandra] abstract class BaseRelationImpl(
+    tableRef: TableRef,
+    connector: CassandraConnector,
+    readConf: ReadConf,
+    writeConf: WriteConf,
+    userSpecifiedSchema: Option[StructType],
+    override val sqlContext: SQLContext)
+  extends BaseRelation
+  with InsertableRelation
+  with Serializable
+  with Logging {
+
+  protected[this] val tableDef =
+    sqlContext.getCassandraSchema(tableRef.cluster.getOrElse("default"))
+      .keyspaceByName(tableRef.keyspace).tableByName(tableRef.table)
+
+  override def schema: StructType = {
+    userSpecifiedSchema.getOrElse(StructType(tableDef.allColumns.map(toStructField)))
+  }
+
+  implicit val cassandraConnector = connector
+  implicit val readconf = readConf
+  protected[this] val baseRdd =
+    sqlContext.sparkContext.cassandraTable[CassandraSQLRow](tableRef.keyspace, tableRef.table)
+
+  override def insert(data: DataFrame, overwrite: Boolean): Unit = {
+    if (overwrite) {
```

Comment:
I'm not sure this is semantically what a C* user would expect.  I think we should not allow an insert to do a truncate in any case. I think that still follows the letter of the interface definition since overwriting will still over write the data of the "relation" even if it doesn't overwrite other existing data ...

> A BaseRelation that can be used to insert data into it through the insert method.
> If overwrite in insert method is true, the old data in the relation should be overwritten with
> the new data. If overwrite in insert method is false, the new data should be appended.